### ISSUE TITLE
The right way to use SM for entity state management

### ISSUE THREAD

**Issue by Pazus** [spring-projects/spring-statemachine#453]
Hi

I've searched web a lot but I'm still confused and haven't found the right implementation of the scenario using SM.
Let's say I have an `Order` entity with `orderState` attribut (Enum). The entity is managed by Spring-data jpa repository. Let's say I have a state diagram `NEW->FORMED->PAID->SHIPPED`. I have a web service for creating of many orders every minut, multithreaded environment.

What is the right implementation? I tried to create a state machin by extending `EnumStateMachineConfigurerAdapter`, and implementing `EnumPersistStateMachineHandler` class (based on `PersistStateMachineHandler` but working with Enums). Then I create a service and a listener, one puts `order` entity into message headers, the other updates the entity's state. But the `EnumPersistStateMachineHandler` (following `PersistStateMachineHandler` approach) executes `stateMachine.stop();` `stateMachine.start();` that doesn't seem like the good solution for multithreaded environment, right? This makes me confused...

Could you provide a link to a guide covering the problem and describing the correct approach fo such use case.
Thanks.

**Comment by jvalkeal** [comment]
Yeah, current facilities and samples we have assumes that you stop a machine. Things get a little awkward [Something I posted to SO](https://stackoverflow.com/questions/47610657/spring-state-machine-jpa-persistence).

I'm finally trying to make this thing better in `1.2.x`. branch and it will be merged to `2.x` when `1.2.8` is released. From that work looks like you also want something I'm implementing in [DefaultStateMachineService](https://github.com/spring-projects/spring-statemachine/blob/1.2.x/spring-statemachine-core/src/main/java/org/springframework/statemachine/service/DefaultStateMachineService.java).

If you're interested to see a bit more, this [spring-cloud-skipper PR 367](https://github.com/spring-cloud/spring-cloud-skipper/pull/367) uses these snapshots and  essentially in that PR use like this [SkipperStateMachineService](https://github.com/jvalkeal/spring-cloud-skipper/blob/4ce69d7eaf3109d8c9ccb2c3225a2d7d67edb430/spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/statemachine/SkipperStateMachineService.java#L155).

This may not help you right now but wanted to mention that some better stuff is coming. Essentially we'd allow OOB persistence via interceptor which goes back to DB via [JPA](https://github.com/spring-projects/spring-statemachine/blob/1.2.x/spring-statemachine-data/jpa/src/main/java/org/springframework/statemachine/data/jpa/JpaPersistingStateMachineInterceptor.java) without stopping/starting a machine.

`datajpapersist` sample I mentioned in SO post uses machine factory and i.e. in your case you'd have machine per entity.

This might also be a good moment to chime in how well these new concepts work with other people.

**Comment by Pazus** [comment]
I believe that missing such basic and important capabilities like managing entities states greatly shrinks the adoption of the project.

As an enchantment request: I like the approach another library implemented when one can annotate an Entity attribute with `State` annotation and te FSM knows which attribute to use as state storage and it integrates with spring-data to manipulate the attribute’s value. Very nice and convenient config. I think using such approach and some genetics you can come up with configuration where user has to only define the state machine (states, transitions etc) and set the entity to manage. Single class config and that’s it! It’d be awesome wouldn’t it?

**Comment by jvalkeal** [comment]
That sounds like a very nice idea. Is this "another library" something publicly available somewhere? I'd very much like to see concepts around it.

**Comment by Pazus** [comment]
http://www.statefulj.org

**Comment by CayoSouza** [comment]
@Pazus @jvalkeal any updates on this topic?
I have the very same requirement and am kinda lost with docs and wether SpringSM fits or not this scenario.
